guiclient/cashReceipt.cpp

```
@@ -973,6 +982,80 @@ bool cashReceipt::save(bool partial)
   return true;
 }
 
+bool cashReceipt::postReceipt()
+{
+  XSqlQuery cashPost;
+  bool changeDate = false;
+  QDate newDate = QDate();
+  
+  if (_privileges->check("ChangeCashRecvPostDate"))
+  {
+    getGLDistDate newdlg(this, "", true);
+    newdlg.sSetDefaultLit(tr("Distribution Date"));
+    if (newdlg.exec() == XDialog::Accepted)
+    {
+      newDate = newdlg.date();
+      changeDate = (newDate.isValid());
+    }
+    else
+      return false;
+  }
+  
+  int journalNumber = -1;
+
+  XSqlQuery tx;
+  tx.exec("BEGIN;");
+  cashPost.exec("SELECT fetchJournalNumber('C/R') AS journalnumber;");
+  if (cashPost.first())
+    journalNumber = cashPost.value("journalnumber").toInt();
+  else if (ErrorReporter::error(QtCriticalMsg, this, tr("Error Posting Cash Receipt Entry"),
+                                cashPost, __FILE__, __LINE__))
+  {
+    return false;
+  }
+
+  XSqlQuery setDate;
+  setDate.prepare("UPDATE cashrcpt SET cashrcpt_distdate=:distdate,"
+                  "                    cashrcpt_applydate=CASE WHEN (cashrcpt_applydate < :distdate) THEN :distdate"
+                  "                                            ELSE cashrcpt_applydate END "
+                  "WHERE cashrcpt_id=:cashrcpt_id;");
+   
+  if (changeDate)
+  {
+    setDate.bindValue(":distdate",    newDate);
+    setDate.bindValue(":cashrcpt_id", _cashrcptid);
+    setDate.exec();
+    ErrorReporter::error(QtCriticalMsg, this, tr("Changing Dist. Date"),
```

This can open a dialog waiting for user input during a transaction, which is generally a bad idea. In addition, the transaction tries to continue after a failure instead of rolling back and returning control to the user.